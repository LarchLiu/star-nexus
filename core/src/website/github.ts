import type { FetchError, FetchWebsite, GithubMeta, NotThrowError, SupabasePicBedRes } from '../types'
import { GITHUB_DOMAIN, GITHUB_RAW_URL, GITHUB_REPOS_API, PICTURE_BED_URL, USER_AGENT } from '../const'
import { fetchGet } from '../utils'

async function getGithubInfo(url: string, picBed?: string, header: Record<string, string> = { 'User-Agent': USER_AGENT }): Promise<FetchWebsite> {
  let title = ''
  let content = ''
  const githubMeta: GithubMeta = { domain: GITHUB_DOMAIN, website: 'Github' }
  const regexGithubRepo = /https:\/\/github.com\/([^\/]*\/[^\/]*)/g // match github.com/user/repo/
  const githubRepoMatch = regexGithubRepo.exec(url)
  const githubRepo = githubRepoMatch ? githubRepoMatch[1] : ''

  try {
    if (githubRepo) {
      // fetch repo info
      const repoJson = await fetchGet<any>(`${GITHUB_REPOS_API}/${githubRepo}`, header)
      // fetch languages
      const languagesJson = await fetchGet(`${GITHUB_REPOS_API}/${githubRepo}/languages`, header)
      // fetch readme
      const readmeRes = await fetchGet<string>(`${GITHUB_RAW_URL}/${githubRepo}/${repoJson.default_branch}/README.md`, header, undefined, false)
      const readme = (readmeRes as NotThrowError).error ? '' : readmeRes

      const description = repoJson.description ? repoJson.description.replace(/:\w+:/g, ' ') : ''
      title = repoJson.full_name + (description ? (`: ${description}`) : '')
      url = repoJson.html_url
      const tags = repoJson.topics

      if (tags && tags.length > 0)
        githubMeta.tags = tags

      if (languagesJson)
        githubMeta.languages = Object.keys(languagesJson)

      const imageBaseUrl = picBed || PICTURE_BED_URL
      if (imageBaseUrl) {
        const user = repoJson.owner.login
        const repo = repoJson.name
        const imageUrl = `${imageBaseUrl}?username=${user}&reponame=${repo}&stargazers_count=${repoJson.stargazers_count}&language=${repoJson.language}&issues=${repoJson.open_issues_count}&forks=${repoJson.forks_count}&description=${description}`
        const imageJson = await fetchGet<SupabasePicBedRes>(imageUrl)
        githubMeta.socialPreview = (imageJson as SupabasePicBedRes).url
      }

      content = `${title}\n\n${readme}`
      if (content.length > 1000) {
        content = content.substring(0, 1000)
        content += '...'
      }
    }
    else {
      return { error: 'Github error: Not supported website.' }
    }
    return { data: { title, url, content, meta: githubMeta } }
  }
  catch (error) {
    return { error: `Github error: ${error as FetchError}` }
  }
}

export { getGithubInfo }
