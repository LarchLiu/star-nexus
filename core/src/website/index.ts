import { websiteLoader } from '../const'
import type { FetchWebsite } from '../types'
import { getDomain } from '../utils'

async function getWebsiteInfo(url: string, picBed?: string, header?: Record<string, string>): Promise<FetchWebsite> {
  let info: FetchWebsite = {}
  const host = getDomain(url)

  if (websiteLoader[host])
    info = await websiteLoader[host].loader(url, picBed, header)
  else
    info.error = 'StarNexus error: Not supported website.'

  return info
}

export {
  getWebsiteInfo,
}
