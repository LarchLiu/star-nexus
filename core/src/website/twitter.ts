import type { FetchError, FetchWebsite, PicBedRes, TwitterMeta } from '../types'
import { PICTURE_BED_URL, STAR_NEXUS_HUB_API, TWITTER_DOMAIN, USER_AGENT } from '../const'
import { fetchGet, fetchPost } from '../utils'

async function getTwitterInfo(url: string, picBed?: string, header: Record<string, string> = { 'User-Agent': USER_AGENT }): Promise<FetchWebsite> {
  let title = ''
  let content = ''
  const meta: TwitterMeta = { domain: TWITTER_DOMAIN, website: 'Twitter' }
  const regexPath = /https:\/\/twitter.com\/([^\/]*\/status\/[^\?]*)/g // match github.com/user/repo/
  const pathMatch = regexPath.exec(url)
  const path = pathMatch ? pathMatch[1] : ''
  const status = path.split('/')[2]

  try {
    if (path) {
      // fetch tweets info
      const resJson = await fetchGet<any>(`${STAR_NEXUS_HUB_API}/twitter/tweet/${path}/original=true`, header)
      title = resJson.title

      const tweets = resJson.item as any[]
      const tweet = tweets.find(({ id_str }) => id_str === status)
      tweets.forEach((t) => {
        let fullText = t.full_text
        if (t.entities?.urls) {
          t.entities.urls.forEach((e: any) => {
            fullText = fullText.replace(e.url, e.expanded_url)
          })
        }
        if (t.quoted_status) {
          let quotedFullText = t.quoted_status.full_text
          if (t.quoted_status.entities?.urls) {
            t.quoted_status.entities.urls.forEach((e: any) => {
              quotedFullText = quotedFullText.replace(e.url, e.expanded_url)
            })
          }

          content += `${t.quoted_status.user.name}: ${quotedFullText}\n${t.user.name}: ${fullText}\n`
        }
        else { content += `${t.user.name}: ${fullText}\n` }
      })

      if (tweet) {
        const imageBaseUrl = picBed || PICTURE_BED_URL
        if (imageBaseUrl) {
          const user = tweet.user
          const name = user.name
          const screenName = user.screen_name
          const avator = user.profile_image_url_https
          const content = user.full_text
          const body = {
            name, screenName, avator, content, status,
          }
          const imageUrl = `${imageBaseUrl}/twitter`
          const imageJson = await fetchPost<PicBedRes>(imageUrl, { ...header, 'Content-Type': 'application/json' }, body)
          meta.cover = (imageJson as PicBedRes).url
        }
      }
    }
    else {
      return { error: 'Twitter error: Not supported website.' }
    }
    return { data: { title, url, content, meta } }
  }
  catch (error) {
    return { error: `Twitter error: ${error as FetchError}` }
  }
}

export { getTwitterInfo }
