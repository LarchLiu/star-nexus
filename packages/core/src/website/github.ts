import type { FetchError, FetchWebsite, GithubMeta, LoaderUrls, NotThrowError, PicBedRes } from '../types'
import { GITHUB_DOMAIN, GITHUB_RAW_URL, GITHUB_REPOS_API, PICTURE_BED_URL, USER_AGENT } from '../const'
import { fetchGet, fetchPost } from '../utils'

async function getGithubInfo(urls: LoaderUrls, header: Record<string, string> = { 'User-Agent': USER_AGENT }): Promise<FetchWebsite> {
  let title = ''
  let content = ''
  let url = urls.webUrl
  const githubMeta: GithubMeta = { domain: GITHUB_DOMAIN, website: 'Github' }
  const regexGithubRepo = /https:\/\/github.com\/([^\/]*\/[^\/]*)/g // match github.com/user/repo/
  const githubRepoMatch = regexGithubRepo.exec(url)
  const githubRepo = githubRepoMatch ? githubRepoMatch[1] : ''

  try {
    if (githubRepo) {
      // fetch repo info
      const repoJson = await fetchGet<any>(`${GITHUB_REPOS_API}/${githubRepo}`, header)
      // fetch languages
      const languagesJson = await fetchGet(`${GITHUB_REPOS_API}/${githubRepo}/languages`, header)
      // fetch readme
      const readmeRes = await fetchGet<string>(`${GITHUB_RAW_URL}/${githubRepo}/${repoJson.default_branch}/README.md`, header, undefined, false)
      const readme = (readmeRes as NotThrowError).error ? '' : readmeRes

      const description = repoJson.description ? repoJson.description.replace(/:\w+:/g, ' ') : ''
      title = repoJson.full_name + (description ? (`: ${description}`) : '')
      url = repoJson.html_url
      const tags = repoJson.topics

      if (tags && tags.length > 0)
        githubMeta.tags = tags

      if (languagesJson)
        githubMeta.languages = Object.keys(languagesJson)

      const imageBaseUrl = urls.picBed || PICTURE_BED_URL
      if (imageBaseUrl) {
        const user = repoJson.owner.login
        const repo = repoJson.name
        const body = {
          username: user,
          reponame: repo,
          stargazers_count: repoJson.stargazers_count,
          language: repoJson.language,
          issues: repoJson.open_issues_count,
          forks: repoJson.forks_count,
          description,
        }
        const imageUrl = `${imageBaseUrl}/github`
        const imageJson = await fetchPost<PicBedRes>(imageUrl, {
          ...header,
          'Content-Type': 'application/json',
        }, body)

        githubMeta.cover = (imageJson as PicBedRes).url
      }

      content = `${title}\n\n${readme}`
    }
    else {
      return { error: 'Github error: Not supported website.' }
    }
    return { data: { title, url, content, meta: githubMeta } }
  }
  catch (error) {
    return { error: `Github error: ${error as FetchError}` }
  }
}

export { getGithubInfo }
