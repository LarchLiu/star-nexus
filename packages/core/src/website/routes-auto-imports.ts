// Auto Generated by '@starnexus/import-routes'
import { router as githubcom } from './github.com/router'
import { pathInfo as githubcomRepo } from './github.com/paths/repo'
import { router as twittercom } from './twitter.com/router'
import { pathInfo as twittercomTweet } from './twitter.com/paths/tweet'

githubcom.paths = [
  githubcomRepo,
].sort((a, b) => (b.sequence ?? 0) - (a.sequence ?? 0))
twittercom.paths = [
  twittercomTweet,
].sort((a, b) => (b.sequence ?? 0) - (a.sequence ?? 0))

export const routes = {
  'github.com': githubcom,
  'twitter.com': twittercom,
}
