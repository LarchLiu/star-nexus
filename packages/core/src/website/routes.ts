import type { Routes } from '../types'
import github from './github.com/router'
import twitter from './twitter.com/router'

const routes: Routes = {
  'github.com': github,
  'twitter.com': twitter,
}

export default routes
