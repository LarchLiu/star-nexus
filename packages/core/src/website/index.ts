import type { FetchWebsite, LoaderUrls, Routes } from '../types'
import { getDomain } from '../utils'
import { routes } from './routes-auto-imports'

export async function getWebsiteInfo(urls: LoaderUrls, header?: Record<string, string>, otherRoutes: Routes = {}): Promise<FetchWebsite> {
  let info: FetchWebsite = { error: 'StarNexus error: Not supported website.' }
  const domain = getDomain(urls.webUrl)
  const allRoutes: Routes = { ...routes, ...otherRoutes }

  if (allRoutes[domain]) {
    const router = allRoutes[domain]
    if (router.paths) {
      for (let i = 0; i < router.paths.length; i++) {
        const pathInfo = router.paths[i]
        const loaderUrls = pathInfo.filter(urls)
        if (loaderUrls) {
          info = await pathInfo.loader(loaderUrls, header)
          if (info.data) {
            info.data.meta.domain = domain
            info.data.meta.website = router.name
          }

          return info
        }
      }
    }
    return { error: `${router.name} error: Not supported website.` }
  }

  return info
}
