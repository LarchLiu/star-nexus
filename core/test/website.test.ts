import { describe, expect, test } from 'vitest'
import { getWebsiteInfo } from '~/website'

describe('get website info', () => {
  test('github info', async () => {
    const res = await getWebsiteInfo('https://github.com/LarchLiu/star-nexus/something')
    expect((res.data?.title)?.includes('star-nexus')).toBeTruthy()
    expect(res.data?.url).toBe('https://github.com/LarchLiu/star-nexus')
  })
  test('github repo path error', async () => {
    const res = await getWebsiteInfo('https://github.com/LarchLiu')
    expect(res.error).toMatchSnapshot()
  })
  test('website domain error', async () => {
    const res = await getWebsiteInfo('https://domain-error.com/xxx/yyy')
    expect(res.error).toMatchSnapshot()
  })
})
