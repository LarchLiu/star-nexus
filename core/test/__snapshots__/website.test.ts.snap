// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get website info > github repo path error 1`] = `"Github error: Not supported website."`;

exports[`get website info > website domain error 1`] = `"StarNexus error: Not supported website."`;
